src/hooks/useParams.ts

```diff
@@ -0,0 +1,54 @@
+import { RouteContext } from '../contexts';
+import { Params } from '@remix-run/router';
+import { useThrottledContext } from './useThrottledContext';
+import { ModalWithRoot } from '../type';
+import { invariant } from '../utils/utils';
+import { useHistoryManager } from './useHistoryManager';
+
+type RequireOnlyOne<T, Keys extends keyof T = keyof T> = Pick<T, Exclude<keyof T, Keys>> &
+  { [K in Keys]-?: Required<Pick<T, K>> & Partial<Record<Exclude<Keys, K>, undefined>> }[Keys];
+
+type AnimatedParts = Pick<ModalWithRoot, 'tab' | 'panel' | 'modal'>;
+type NavId = RequireOnlyOne<AnimatedParts>;
+
+export function useParams<T extends string = string>(id?: NavId): Params<T> | undefined {
+  const [routeContext] = useThrottledContext(RouteContext);
+  const historyManager = useHistoryManager();
+
+  invariant(
+    routeContext,
+    'You can not use useParams hook outside of RouteContext. Make sure calling it inside RouterProvider.',
+  );
+
+  const historyStack = historyManager.getHistory();
+  const currentPosition = historyManager.getCurrentPosition();
+  const previousRecord = currentPosition ? historyStack[currentPosition - 1] : undefined;
+
+  let recordMatchesId = false;
+  let currentRouteDoesNotMatchId = false;
+
+  /**
+   * Checks if the provided `id` matches the previous history record and differs from the current route,
+   * ensuring correct swipe-back animation and proper modal closing.
+   */
+  if (id) {
+    const idKeys = Object.keys(id) as (keyof NavId)[];
+
+    if (previousRecord) {
+      recordMatchesId = idKeys.every((key) => id[key] === previousRecord[key]);
```

Comment:
Лучше заменить на some с инверсией условия. тогда не придется проходить по всем ключам

```suggestion
      recordMatchesId = !idKeys.some((key) => id[key] !== previousRecord[key]);
```